### ISSUE TITLE
`Iterable[str]` unpacking into `Literal` is valid from 3.11 on

### ISSUE THREAD

**Issue by karolyi** [microsoft/pyright#9194]
Hey, 

not sure this is a 'bug' or a 'feature request'. Anyways:

Python3.11 supports the unpacking of `Iterable[str]` into a `Literal`. This is useful for generating literals from variables for type checking.

Small example on 3.11:
```python
Python 3.11.9 (main, Aug  1 2024, 12:59:41) [GCC 14.1.1 20240522] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> typing.Literal[*('1', '2')]
typing.Literal['1', '2']
```

3.10 doesn't support it:
```python
Python 3.10.14 (main, Apr 29 2024, 09:49:44) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> typing.Literal[*('1', '2')]
  File "<stdin>", line 1
    typing.Literal[*('1', '2')]
                   ^
SyntaxError: invalid syntax

```

Pyright doesn't seem to support it. Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.11&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2AUNQB5QC8hxZVA2gFQAUA5AIy8ANFF4AmXgEoAutSA)

```python
from typing import Literal

x = Literal[*('1', '2')]

```

**Comment by karolyi** [comment]
Expanding on why this is important, with a more precise example:

```python
Python 3.11.9 (main, Aug  1 2024, 12:59:41) [GCC 14.1.1 20240522] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Literal
>>> from inspect import signature, Parameter
>>> 
>>> valid_values = list(signature(Parameter).parameters)
>>> 
>>> ParamAttribute = Literal[*valid_values]
>>> assert ParamAttribute == Literal['name', 'kind', 'default', 'annotation']
>>> ParamAttribute
typing.Literal['name', 'kind', 'default', 'annotation']
```

**Comment by karolyi** [comment]
Upon diving more into it, it seems that py3.10 also supports the creation of literals from (proper) variables:

```python
Python 3.10.14 (main, Apr 29 2024, 09:49:44) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Literal
>>> from typing_extensions import reveal_type
>>> x = '1'
>>> y = '2'
>>> Literal[(x, y)]
typing.Literal['1', '2']
>>> xy = Literal[(x, y)]
>>> reveal_type(xy)
Runtime type is '_LiteralGenericAlias'
typing.Literal['1', '2']

>>> z = 1
>>> xyz = Literal[(x, y, z)]
>>> reveal_type(xyz)
Runtime type is '_LiteralGenericAlias'
typing.Literal['1', '2', 1]
```

Pyright doesn't seem to support it:
![Image](https://github.com/user-attachments/assets/bc55739e-d8fd-49cb-adfe-5b6431fd6d15)

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.385.

**Comment by karolyi** [comment]
I was only trying to achieve what Python already does; you can test the output in an actual python REPL, it will give you the values I specified.

If you don't intend to achieve similar functionality to that of python, that's fine, but if think you'll run into more problems regarding this in the future. Not from me by any means, but I can envision other people complaining about the same exact problem.

Anyways, thanks for clearing up your stance.

**Comment by erictraut** [comment]
> it will give you the values I specified

You are conflating *values* and *types*. At runtime, the interpreter evaluates the *value* of an expression. A static type checker evaluates the *type* of an expression.

You're also confusing the representation of a value (the human-readable string returned by `__repr__` that you see printed on the REPL) with the actual class that an object derives from. Type checkers are concerned with the latter. In this case, the REPL is printing `Literal['1', '2']`, but the object is actually an instance of `typing._LiteralGenericAlias`.

The *type* of the symbol `typing.Literal` is defined as `_SpecialForm` in `typing.pyi`. `_SpecialForm` is a placeholder that means "this is an implementation detail of the type system". At runtime, the expression `typing.Literal` actually evaluates to an object that is an instance of the undocumented class `typing._LiteralSpecialForm`, and the expression `Literal['1', '2']` evaluates to an instance of an undocumented class named `typing._LiteralGenericAlias`. Since these are undocumented and have no definitions supplied in the stdlib typeshed stubs, pyright simply displays their types as `Literal`. That's what you see when using the `reveal_type` call.

For comparison, mypy reveals `Any` or `_SpecialForm` in your code sample. Both pyright and mypy are working in conformance with the typing spec since the symbol `Literal` isn't defined when used in within a value expression. The [typing spec defines its meaning](https://typing.readthedocs.io/en/latest/spec/literal.html#literals) only in the case where it's used in a type expression.

**Comment by karolyi** [comment]
Ah, there lies the crux of the matter, thanks.

I really don't want to keep bugging you with this, wasting your time. Pyright (and its derivatives) is already the fastest type checker that exists out there, I can live with these nuances. It's just I didn't dig this deep into it to realize what you've outlined here.

Thanks for the help again!
